--- Day 5: Cafeteria ---
As the forklifts break through the wall, the Elves are delighted to discover that there was a cafeteria on the other side after all.

You can hear a commotion coming from the kitchen. "At this rate, we won't have any time left to put the wreaths up in the dining hall!" Resolute in your quest, you investigate.

"If only we hadn't switched to the new inventory management system right before Christmas!" another Elf exclaims. You ask what's going on.

The Elves in the kitchen explain the situation: because of their complicated new inventory management system, they can't figure out which of their ingredients are fresh and which are spoiled. When you ask how it works, they give you a copy of their database (your puzzle input).

The database operates on ingredient IDs. It consists of a list of fresh ingredient ID ranges, a blank line, and a list of available ingredient IDs. For example:

3-5
10-14
16-20
12-18

1
5
8
11
17
32
The fresh ID ranges are inclusive: the range 3-5 means that ingredient IDs 3, 4, and 5 are all fresh. The ranges can also overlap; an ingredient ID is fresh if it is in any range.

The Elves are trying to determine which of the available ingredient IDs are fresh. In this example, this is done as follows:

Ingredient ID 1 is spoiled because it does not fall into any range.
Ingredient ID 5 is fresh because it falls into range 3-5.
Ingredient ID 8 is spoiled.
Ingredient ID 11 is fresh because it falls into range 10-14.
Ingredient ID 17 is fresh because it falls into range 16-20 as well as range 12-18.
Ingredient ID 32 is spoiled.
So, in this example, 3 of the available ingredient IDs are fresh.

Process the database file from the new inventory management system. How many of the available ingredient IDs are fresh?

In [ ]:
from aocd import get_data, submit

current_day is only available in December (EST)


In [ ]:
data = get_data(year = 2025, day = 5)

In [ ]:
data_ls = data.splitlines()

In [ ]:
data_ls[:5]

['436639876131680-436867778393908',
 '197700960860637-198893061877806',
 '255753644093984-256141049798709',
 '2351348670325-2617031498963',
 '132305455060243-136355961131249']

In [ ]:
sum(["-" in rng for rng in data_ls])

186

In [ ]:
len(data_ls)

1187

In [ ]:
data_ls[-10]

'389350147844854'

it looks from the example and the data like the input data format has ranges with a "-" in them and ingredient ids as as single int in the same data. can you write a function that splits this out into ranges and ingredients

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Sure! Here's a function that splits the data into ranges and ingredient IDs:

```python
def parse_data(data_ls):
    ranges = []
    ingredients = []
    
    for line in data_ls:
        if '-' in line:
            start, end = map(int, line.split('-'))
            ranges.append((start, end))
        elif line:  # non-empty line without '-'
            ingredients.append(int(line))
    
    return ranges, ingredients
```

This checks each line for a hyphen to distinguish ranges from ingredient IDs. Try running it on your data and see what you get!

In [ ]:
def parse_data(data_ls):
    ranges = []
    ingredients = []
    
    for line in data_ls:
        if '-' in line:
            start, end = map(int, line.split('-'))
            ranges.append((start, end))
        elif line:  # non-empty line without '-'
            ingredients.append(int(line))
    
    return ranges, ingredients

In [ ]:
example_ls = """3-5
10-14
16-20
12-18

1
5
8
11
17
32""".splitlines()

example_ranges, example_ingredients = parse_data(example_ls)

In [ ]:
data_ranges, data_ingredients = parse_data(data_ls)

can you write code that for any list of ingredients loops over the ranges checking if ingredient between min and max, marking in a bool list, and at the end checking if any true

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's a function that does that:

```python
def is_fresh(ingredient, ranges):
    return any(start <= ingredient <= end for start, end in ranges)
```

This checks if the ingredient falls within any of the ranges. You can then use it to count fresh ingredients:

```python
fresh_count = sum(is_fresh(ing, data_ranges) for ing in data_ingredients)
```

Want to try it out on the example first to verify it works?

In [ ]:
def check_fresh(ingredients, ranges):
    fresh = []
    for ingredient in ingredients:
        is_fresh = any(start <= ingredient <= end for start, end in ranges)
        fresh.append(is_fresh)
    return fresh

In [ ]:
def pt1(data_ls):
    ranges, ingredients = parse_data(data_ls)
    fresh = check_fresh(ingredients, ranges)
    return sum(fresh)

In [ ]:
pt1(example_ls)

3

In [ ]:
pt1_answer = pt1(data_ls)
pt1_answer

865

In [ ]:
submit(pt1_answer, part = "a", year = 2025, day = 5)

Part a already solved with same answer: 865


--- Day 5: Cafeteria ---
As the forklifts break through the wall, the Elves are delighted to discover that there was a cafeteria on the other side after all.

You can hear a commotion coming from the kitchen. "At this rate, we won't have any time left to put the wreaths up in the dining hall!" Resolute in your quest, you investigate.

"If only we hadn't switched to the new inventory management system right before Christmas!" another Elf exclaims. You ask what's going on.

The Elves in the kitchen explain the situation: because of their complicated new inventory management system, they can't figure out which of their ingredients are fresh and which are spoiled. When you ask how it works, they give you a copy of their database (your puzzle input).

The database operates on ingredient IDs. It consists of a list of fresh ingredient ID ranges, a blank line, and a list of available ingredient IDs. For example:

3-5
10-14
16-20
12-18

1
5
8
11
17
32
The fresh ID ranges are inclusive: the range 3-5 means that ingredient IDs 3, 4, and 5 are all fresh. The ranges can also overlap; an ingredient ID is fresh if it is in any range.

The Elves are trying to determine which of the available ingredient IDs are fresh. In this example, this is done as follows:

Ingredient ID 1 is spoiled because it does not fall into any range.
Ingredient ID 5 is fresh because it falls into range 3-5.
Ingredient ID 8 is spoiled.
Ingredient ID 11 is fresh because it falls into range 10-14.
Ingredient ID 17 is fresh because it falls into range 16-20 as well as range 12-18.
Ingredient ID 32 is spoiled.
So, in this example, 3 of the available ingredient IDs are fresh.

Process the database file from the new inventory management system. How many of the available ingredient IDs are fresh?

Your puzzle answer was 865.

The first half of this puzzle is complete! It provides one gold star: *

--- Part Two ---
The Elves start bringing their spoiled inventory to the trash chute at the back of the kitchen.

So that they can stop bugging you when they get new inventory, the Elves would like to know all of the IDs that the fresh ingredient ID ranges consider to be fresh. An ingredient ID is still considered fresh if it is in any range.

Now, the second section of the database (the available ingredient IDs) is irrelevant. Here are the fresh ingredient ID ranges from the above example:

3-5
10-14
16-20
12-18
The ingredient IDs that these ranges consider to be fresh are 3, 4, 5, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, and 20. So, in this example, the fresh ingredient ID ranges consider a total of 14 ingredient IDs to be fresh.

Process the database file again. How many ingredient IDs are considered to be fresh according to the fresh ingredient ID ranges?

hmmmm. should probably assume a strait loop wont work. 

can you write code that will try a loop (i.e. iter over ranges adding a to a set), but timeouts after 30 secs

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Here's code that tries the loop approach with a timeout:

```python
import signal

def timeout_handler(signum, frame):
    raise TimeoutError("Operation timed out")

def pt2_with_timeout(ranges, timeout_seconds=30):
    signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(timeout_seconds)
    
    try:
        fresh_ids = set()
        for start, end in ranges:
            for i in range(start, end + 1):
                fresh_ids.add(i)
        signal.alarm(0)  # Cancel the alarm
        return len(fresh_ids)
    except TimeoutError:
        signal.alarm(0)
        return None  # Indicates timeout
```

Try running `pt2_with_timeout(data_ranges)` and see if it times out. If it does, we'll need a smarter approach!

In [ ]:
import signal

def timeout_handler(signum, frame):
    raise TimeoutError("Operation timed out")

def pt2_with_timeout(ranges, timeout_seconds=30):
    signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(timeout_seconds)
    
    try:
        fresh_ids = set()
        for start, end in ranges:
            for i in range(start, end + 1):
                fresh_ids.add(i)
        signal.alarm(0)  # Cancel the alarm
        return len(fresh_ids)
    except TimeoutError:
        signal.alarm(0)
        return None  # Indicates timeout

In [ ]:
pt2_with_timeout(data_ranges)

yeah thought that wouldn't work. 

how else can I approach this?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Good instinct! Since the ranges are huge, we need to avoid iterating through individual IDs.

The key insight: instead of expanding ranges into individual numbers, work with the ranges themselves. You need to:

1. Merge overlapping ranges into non-overlapping ones
2. Sum up the lengths of the merged ranges

For example, ranges `3-5` and `10-14` don't overlap, so they contribute `3 + 5 = 8` IDs. But `10-14` and `12-18` overlap, so they merge into `10-18` which contributes `9` IDs.

Want to try implementing a function to merge overlapping ranges? Or would you like a hint on the algorithm?

hmm. I'm struggling here. does this involve a loop of all ranges to check if the min / max falls within other ranges?

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Not quite! You don't need to check every range against every other range individually.

The trick is to **sort the ranges first** by their start positions. Then you can merge them in a single pass through the sorted list.

Here's the idea:
- Sort ranges by start value
- Keep track of the "current merged range"
- For each new range, check if it overlaps with or extends the current one
- If yes, extend the current range; if no, save it and start a new one

Does that make sense? Want to try coding the merge logic, or would you like me to show you a small example of how two ranges would merge?

yeah that makes sense. can you start sorting the ranges

##### 🤖Reply🤖<!-- SOLVEIT_SEPARATOR_7f3a9b2c -->

Sure! Here's how to sort the ranges by their start value:

```python
sorted_ranges = sorted(data_ranges, key=lambda r: r[0])
```

This sorts the list of tuples by the first element (the start of each range). Try running that and take a look at the first few sorted ranges to see how they're ordered!

In [ ]:
sorted_ranges = sorted(data_ranges, key=lambda r: r[0])

In [ ]:

int(sorted_ranges[-1][1])

561397097410102

In [ ]:
sorted_ranges[:5]


[(77643299597, 926402139635),
 (410879657749, 1138886722630),
 (1138886722630, 1854190426883),
 (1138886722630, 1854190426883),
 (1854190426883, 2617031498963)]

In [ ]:
def pt2(ranges):
    sorted_ranges = sorted(ranges, key=lambda r: r[0])

    total = 0
    current_min, current_max = sorted_ranges[0]
    for i, (min_, max_) in enumerate(sorted_ranges):
        if min_ > current_max:
            total += (current_max - current_min) + 1
            current_min = min_
            current_max = max_

        if max_ > current_max:
            current_max = max_

    # final add
    total += (current_max - current_min) + 1

    return total

In [ ]:
pt2(example_ranges)

14

In [ ]:
pt2_answer = pt2(data_ranges)

In [ ]:
submit(pt2_answer, part = "b", year = 2025, day = 5)

That's the right answer!  You are one gold star closer to decorating the North Pole.You have completed Day 5! You can [Shareon
  Bluesky
Twitter
Mastodon] this victory or [Return to Your Advent Calendar].
